inst/include/iterative_solver.h

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023 'finalsize' authors. See repository licence in LICENSE.md.
+#ifndef INST_INCLUDE_ITERATIVE_SOLVER_H_
+#define INST_INCLUDE_ITERATIVE_SOLVER_H_
+
+// clang-format off
+#include <Rcpp.h>
+#include <RcppEigen.h>
+
+#include <algorithm>
+// clang-format on
+
+// add to namespace finalsize
+namespace finalsize {
+
+/// function for iterative solver
+// taken from Edwin van Leeuwen at
+// https://gitlab.com/epidemics-r/code_snippets/feature/newton_solver/include/finalsize.hpp
+
+/// @brief Solve an equation using the Newton method with tolerance checking
+/// @param contact_matrix Social contact matrix. Entry \eqn{mm_{ij}} gives
+/// the average number of contacts in group \eqn{i} reported by participants
+/// in group \eqn{j}.
+/// @param demography_vector Demography vector. Entry \eqn{pp_{i}} gives
+/// proportion of total population in group \eqn{i}. Normalised if needed.
+/// @param susceptibility A matrix giving the susceptibility of individuals
+/// in demographic group \eqn{i} and risk group \eqn{j}.
+/// @param iterations Number of solver iterations. Defaults to 10,000.
+/// @param tolerance Solver error tolerance. Solving ends when the error
+/// drops below this tolerance. Defaults to set `1e-6`. Larger tolerance
+/// values are likely to lead to inaccurate final size estimates.
+/// @param step_rate The solver step rate. Defaults to 1.9 as a value found to
+/// work well.
+/// @param adapt_step Boolean, whether the solver step rate should be changed
+/// based on the solver error. Defaults to TRUE.
+/// @return A vector of final sizes.
+inline Eigen::ArrayXd solve_final_size_iterative(
+    const Eigen::MatrixXd &contact_matrix,
+    const Eigen::VectorXd &demography_vector,
+    const Eigen::VectorXd &susceptibility, const int iterations = 10000,
+    const double tolerance = 1e-6, double step_rate = 1.9,
+    const bool adapt_step = true) {
+  // count number of demography groups
+  int nDim = demography_vector.size();
+
+  Eigen::ArrayXi zeros(nDim);
+  zeros.fill(0);
+
+  Eigen::ArrayXd epi_final_size(nDim);  // prev in settings struct
+  epi_final_size.fill(0.5);
```

Comment:
I don't think so. For example, the simple case of finding the final size is included as a function in the tests here as
https://github.com/epiverse-trace/finalsize/blob/2d7c68177f718e917a8853a6d7415f5fb6524318/tests/testthat/test-statistical_correctness.R#L5-L15

The method here also starts with a guess of 0.5 if I've understood correctly, so this might be a standard starting point. Happy for @BlackEdder or @adamkucharski to weigh in.

Reply:
Yes we can get a better starting point based on pre-filled estimates from a non-age structured final size equation. If you solve for R0s in 1:20 (for example) that will give you a place to start for all age groups.

Reply:
Thanks - I haven't really understood that. We are already passing $R_0$ to the solver. Also the snippet I've linked isn't used in the solver code, it's only in the tests - Hugo is referring to the solver code, where this function isn't available (R vs C++).

I could add an 'initial guess' parameter to the solvers, so that it's 0.5 by default, but could also be a vector coming from an earlier solver run on non-age structured populations? This might get a little recursive.